akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -757,11 +786,38 @@ abstract class ShardCoordinator(
               self.tell(GetShardHome(shard), ignoreRef)
             }
           } else {
-            // rebalance not completed, graceful shutdown will be retried
-            gracefulShutdownInProgress -= state.shards(shard)
             clearRebalanceInProgress(shard)
           }
         } else {
+          log.warning("Rebalance shard [{}] didn't complete within [{}].", shard, handOffTimeout.pretty)
+          clearRebalanceInProgress(shard)
+        }
+
+      case ShardShutdownDone(shard, ok) =>
+        rebalanceWorkers -= sender()
+
+        if (ok) {
+          log.debug("Shard [{}] shutdown completed successfully.", shard)
+          if (state.shards.contains(shard)) {
+            update(ShardHomeDeallocated(shard)) { evt =>
+              log.debug("Shard [{}] deallocated after after region shutdown", shard)
+              state = state.updated(evt)
+              clearRebalanceInProgress(shard)
+              allocateShardHomesForRememberEntities()
+              self.tell(GetShardHome(shard), ignoreRef)
+            }
+          } else {
+            clearRebalanceInProgress(shard)
+          }
+        } else {
+          log.warning("Shard [{}] shutdown didn't complete within [{}].", shard, handOffTimeout.pretty)
+          // TODO: we should take that into consideration for future rebalances
+          // the region is trying to shutdown but hangs because of a shard,
+          // we may want to not allocate new shards on this region
+
+          // as a shutdown failed we need to put region back on the shelf
+          // graceful shutdown will be retried
+          gracefulShutdownInProgress -= state.shards(shard)
```

Comment:
here we know that we are dealing with a graceful shutdown, so in case of failure we put the region back to 'active'.

Reply:
I'm not sure it is better to have two separate cases for `ShardShutdownDone` and `RebalanceDone`. They should do pretty much the same aside from some logging. If it's important to adjust the logging here also it's possible to see if it's a rebalance or shutdown by looking at `gracefulShutdownInProgress.contains(state.shards(shard))`